### DISCUSSION TITLE
Set the python working directory as the current latex directory

### DISCUSSION THREAD

**Discussion by jowe19** [tndrle/PyLuaTeX#5] [General]
Is there an option for setting the the python working directory as the current latex directory ? For the moment, I have to modify the `python path` file by file.

**Comment by tndrle** [comment]
How do you define "current latex directory"? Can you show an MWE of your current approach, please?

As an example, if you have the following file in the folder `path/to/file` (placeholder only)
and you run `lualatex -shell-escape file.tex` inside `path/to/file`, then the output
in the document is `path/to/file` (at least on my system). Do you expect something different here?
```latex
\documentclass{article}

\usepackage{pyluatex}

\begin{document}
\begin{python}
import os
print(os.getcwd())
\end{python}
\end{document}
```

**Reply by jowe19** [reply]
Thanks again for replying me. 

- Firstly, with the code

```
\documentclass{article}

\usepackage{pyluatex}

\begin{document}
\begin{python}
import os
print(os.getcwd())
\end{python}
\end{document}
```
I get an error `Missing $ inserted` because my path has an `_`. 

- Secondly, if I put a `.py` called `file_aux.py` in the same directory as my latex file, the code
```
\documentclass{article}

\usepackage{pyluatex}

\begin{document}

\begin{python}
import file_aux
\end{python}
```
gives an error. Whereas the following works fine. 
```
\documentclass{article}

\usepackage{pyluatex}

\begin{document}

\begin{python}
import sys
sys.path.append('path/to/latex_file')
import file_aux
\end{python}
```

**Reply by tndrle** [reply]
Thanks for the details.

For the first point, this should help:
```latex
\documentclass{article}

\usepackage{pyluatex}

\begin{document}

\begin{python}
import sys
import os
with open('output.txt', 'w') as file:
    file.write(' '.join(sys.path) + '\n')
    file.write(os.getcwd() + '\n')
\end{python}

\end{document}
```

Actually, Python's _current working directory_ should be the folder of your LaTeX file.
However, the current working directory is not necessarily in the Python path.
As per the [docs](https://docs.python.org/3/library/sys.html#sys.path):

> As initialized upon program startup, the first item of this list, path[0], is the directory containing the script that was used to invoke the Python interpreter. If the script directory is not available (e.g. if the interpreter is invoked interactively or if the script is read from standard input), path[0] is the empty string, which directs Python to search modules in the current directory first. Notice that the script directory is inserted before the entries inserted as a result of PYTHONPATH.

For PyLuaTeX, the first entry of the path, `path[0]`, is the PyLuaTeX package folder (something like `.../texmf/tex/latex/pyluatex`) because there the Python script that interpretes PyLuaTeX stuff is located.

This explains why your local import does not work, I guess. A simple work around could be `sys.path.insert(0, '')`, but that doesn't change much about your initial question. I'll think about reasonable approaches. At the moment, PyLuaTeX does not offer any related options.

**Comment by tndrle** [comment]
Version 0.4.3 now adds the _folder containing the LaTeX file that was used to invoke LuaLaTeX_ to the Python path automatically. So importing Python files that are located in the same folder as the LaTeX file should be possible now without manually modifying the Python path.

Notice that the _folder containing the LaTeX file_ and __not__ the _current working directory_ (which is the folder you are currently in when you start LuaLaTeX) is added.
This is in the style of the default Python behavior (https://docs.python.org/3/library/sys.html#sys.path):

> As initialized upon program startup, the first item of this list, path[0], is the directory containing the script that was used to invoke the Python interpreter.

Often, both folders coincide of course.

You can disable this behavior with `\usepackage[localimports=false]{pyluatex}`.

Does that solve your original question?

**Comment by jowe19** [comment]
Fixed, it works fine. Thank you for the patch.

**Comment by jowe19** [comment]
Hi, would it be possible to extend this path setting to the inputed latex ? Currently, if I import an other tex file with `\input`, the path of the folder containing the imported file is not added to the python path.

**Reply by tndrle** [reply]
Hi, that would mean pyluatex has to hook into `\input` somehow. This is of course possible (https://tex.stackexchange.com/a/59838), but I want to keep the package as lean as possible. So I will not implement this functionality for now.
